### PR TITLE
Improvement of cookie management

### PR DESCRIPTION
**Pull request by felix2502**
### Checklist

- [x] I have followed (at least) the [PR section of the contributing guide](https://github.com/openkfw/TruBudget/blob/main/CONTRIBUTING.md).
- [x] I fixed all necessary PR warnings
- [x] The commit history is clean
- [x] The E2E tests are passing
- [x] If possible, the issue has been divided into more subtasks
- [x] I did a self review before requesting a review from another team member

### CODE REVIEW

**Review by mayrmartin — CHANGES_REQUESTED** [review]

File: email-notification-service/src/middleware.ts [hunk 1]
```diff
@@ -68,6 +68,9 @@ const verifyJWTToken = (
 };
 
 function getJWTToken(req: Request): string {
+  if (req.cookies.token) {
+      req.headers.authorization = `Bearer ${req.cookies.token}`;
+  }
```
[the remark this answers is not in the document]
Comment (felix2502) [reply email-notification-service/src/middleware.ts:73]: An empty String is set, if there is no token and **verifyJWTToken** will throw the error.

File: api/src/httpd/server.ts [hunk 2]
```diff
@@ -22,10 +22,21 @@ const ajv = new Ajv({
...
 
+  server
+    .register(require("@fastify/cookie"));
+
   server.decorate("authenticate", async (request, reply) => {
     try {
+      const token = request.cookies.token;
+      if (token) {
+        request.headers.authorization = `Bearer ${token}`;
+      }
```
[the remark this answers is not in the document]
Comment (felix2502) [reply api/src/httpd/server.ts:39]: No token will throw the error below. Setting the token every time is for using multiple accounts on the same browser and providing a new token. Token expiration is handled while creating.

[the remark this answers is not in the document]
Comment (mayrmartin) [reply api/src/httpd/server.ts:39]: There is no requirement for multiple accounts in the same browser. Creating a new Token on each request is not good practice.
Further: If the token gets replaced on each request, the UI (in our case Axios) should handle this token change as well.

[the remark this answers is not in the document]
Comment (felix2502) [reply api/src/httpd/server.ts:39]: A token is only created once when the user logs in, and not on every request. On every request it just checks, whether the cookies have a token set or not, if they do, then the headers are set with that token.

File: api/src/user_authenticate.ts [hunk 3]
```diff
@@ -10,6 +10,9 @@ import { AuthToken } from "./service/domain/organization/auth_token";
 import { Group } from "./service/domain/organization/group";
 import { ServiceUser } from "./service/domain/organization/service_user";
 import Joi = require("joi");
+import type { FastifyCookieOptions } from "@fastify/cookie";
+import cookie from "@fastify/cookie";
+import { randomString } from "./service/hash";
```
Comment (Stezido) [inline api/src/user_authenticate.ts:15]: remove unused imports

**Review by Stezido — CHANGES_REQUESTED** [review]
This PR is a breaking change. We only can put it into 3.0.0 with the current changes. Another solution is to still support the old way of storing the JWT token in local storage and remove it with TB 3.0.

**Review by Stezido — APPROVED** [review]
Looks good to me,
I tested the minimal setup using the PRs version for each service.
I also tested it including an old blockchain and export instance (current main).
Good job.

**Comment by Stezido** [comment]
I rebased the branch onto the latest main branch and added a local docker-compose e2e-test fix

**Pull request merged**